cmd/influxd/inspect/report_tsi/report_tsi.go

```diff
@@ -0,0 +1,481 @@
+	if len(r.lowCardinality) > useBitmapN {
+		r.set = tsdb.NewSeriesIDSet()
+		for id := range r.lowCardinality {
+			r.set.AddNoLock(uint64(id))
+		}
+		r.lowCardinality = nil
+	}
+}
+
+func (r *result) merge(other *tsdb.SeriesIDSet) {
+	if r.set == nil {
+		r.set = tsdb.NewSeriesIDSet()
+		for id := range r.lowCardinality {
+			r.set.AddNoLock(uint64(id))
+		}
+		r.lowCardinality = nil
+	}
+	r.set.Merge(other)
+}
+
+type results []*result
+
+func (a results) Len() int           { return len(a) }
+func (a results) Less(i, j int) bool { return a[i].count < a[j].count }
+func (a results) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func (report *reportTSI) printSummaryByMeasurement() error {
+	// Get global set of measurement names across shards.
+	idxs := &tsdb.IndexSet{SeriesFile: report.sfile}
+	for _, idx := range report.shardIdxs {
+		idxs.Indexes = append(idxs.Indexes, idx)
+	}
+
+	mitr, err := idxs.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if mitr == nil {
+		return errors.New("got nil measurement iterator for index set")
+	}
+	defer mitr.Close()
+
+	var name []byte
+	var totalCardinality int64
+	measurements := results{}
+	for name, err = mitr.Next(); err == nil && name != nil; name, err = mitr.Next() {
+		res := &result{name: name}
+		for _, shardCards := range report.cardinalities {
+			other, ok := shardCards[string(name)]
+			if !ok {
+				continue // this shard doesn't have anything for this measurement.
+			}
+
+			if other.short != nil && other.set != nil {
+				panic("cardinality stored incorrectly")
+			}
+
+			if other.short != nil { // low cardinality case
+				res.addShort(other.short)
+			} else if other.set != nil { // High cardinality case
+				res.merge(other.set)
+			}
+
+			// Shard does not have any series for this measurement.
+		}
+
+		// Determine final cardinality and allow intermediate structures to be
+		// GCd.
+		if res.lowCardinality != nil {
+			res.count = int64(len(res.lowCardinality))
+		} else {
+			res.count = int64(res.set.Cardinality())
+		}
+		totalCardinality += res.count
+		res.set = nil
+		res.lowCardinality = nil
+		measurements = append(measurements, res)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// sort measurements by cardinality.
+	sort.Sort(sort.Reverse(measurements))
+
+	if report.topN > 0 {
+		// There may not be "topN" measurement cardinality to sub-slice.
+		n := int(math.Min(float64(report.topN), float64(len(measurements))))
+		measurements = measurements[:n]
+	}
+
+	tw := tabwriter.NewWriter(os.Stdout, 8, 8, 1, '\t', tabwriter.AlignRight)
+
+	fmt.Fprintf(tw, "Summary\nDatabase Path: %s\nCardinality (exact): %d\n\n", report.enginePath, totalCardinality)
+	fmt.Fprint(tw, "Measurement\t\tCardinality (exact)\n\n")
+	for _, res := range measurements {
+		fmt.Fprintf(tw, "%q\t%d\t\n", res.name, res.count)
+	}
+
+	if err := tw.Flush(); err != nil {
+		return err
+	}
+	fmt.Fprint(os.Stdout, "\n\n")
```

Comment:
I would remove this line, I don't think there's value to a bunch of blank lines at the end of the output.